Src/VimApp/VimAppJoinableTaskFactoryProvider.cs

```diff
@@ -0,0 +1,38 @@
+﻿using Microsoft.VisualStudio.Threading;
+using System.ComponentModel.Composition;
+using Vim.UI.Wpf;
+
+namespace VimApp
+{
+    [Export(typeof(IJoinableTaskFactoryProvider))]
+    internal class VimAppJoinableTaskFactoryProvider : IJoinableTaskFactoryProvider
+    {
+        private static readonly object s_lock = new object();
+        private static JoinableTaskContext s_joinableTaskContext;
+
+        private readonly JoinableTaskFactory _joinableTaskFactory;
+
+        [ImportingConstructor]
+        VimAppJoinableTaskFactoryProvider()
+        {
+            // See:
+            // https://github.com/microsoft/vs-threading/blob/master/doc/library_with_jtf.md
+            // https://github.com/microsoft/vs-threading/blob/master/doc/testing_vs.md
+            lock (s_lock)

```

Comment:
Don't think the lock should be necessary here. This MEF should be a singleton so it should only be constructed once.

Reply:
Looked through the docs you linked to and didn't see a `lock` there. Is there a scenario I'm not thinking of? 

---
In reply to: [287919513](https://github.com/VsVim/VsVim/pull/2571#discussion_r287919513) [](ancestors = 287919513)

Reply:
The docs were very insistent that if two or more `JoinableTaskFactory` objects were in use that there would be negative consequences. Although the MEF scenario is predictable, unit tests can construct the provider directly, and if tests were run in parallel we want to be certain that only one gets created.

Admittedly we don't do any of that now, but the lock seemed like an insurance policy that makes it clear that there is one and only one object created per app domain. A similar effect could be achieved by just making it static and then we would rely on static constructor locking. But that scenario creates the object at an unpredictable early time, and given the ominous warnings, this approach ensures it isn't created unless we intend to use it.

TL;DR I was being paranoid.